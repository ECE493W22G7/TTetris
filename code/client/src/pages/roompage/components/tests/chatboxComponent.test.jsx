import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ChatboxComponent from '../chatboxComponent';
import SocketMock from 'socket.io-mock';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import structuredClone from '@ungap/structured-clone';
const { act } = renderer;

expect.extend({ toMatchDiffSnapshot });

let createClient = (id, nick, color) => {
    return {
        "id": id,
        "nickname": nick,
        "color": color
    }
}

let createMessage = (t, nn, msg) => {
    return {
        time: t,
        nickname: nn,
        message: msg
    }
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

describe("ChatboxComponent", () => {
    let socket;
    let history;
    let now = 1649451049840;

    beforeEach(() => {
        history = [];
        history.push(createMessage(now, "player1", "Lorem"));
        now += 1000;
        history.push(createMessage(now, "player2", "Ipsum"));
        now += 60000;
        history.push(createMessage(now, "player3", "Dolor"));
        
        socket = new SocketMock();
        socket.on("getMessage", (roomId, done) => {
            done(history);
        })
    })
    test('renders correctly', async () => {
        let component;
        act(() => {
            component = renderer.create(<ChatboxComponent socket={socket.socketClient} />);
        });

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("should send user input to socket", (done) => {
        let test_msg = "Hello World";

        socket.on("sendMessage", (roomID, msg, clientID) => {
            expect(msg).toEqual(test_msg);
            done()
        })

        render(<ChatboxComponent socket={socket.socketClient} />);
        userEvent.type(screen.getByTestId("send-box-input"), test_msg);
        userEvent.click(screen.getByText("Send"));
    })

    test('should show message when a socket event is recieved', async () => {
        console.log("test2");
        let component;
        act(() => {
            component = renderer.create(<ChatboxComponent socket={socket.socketClient} />);
        });
        let init = component.toJSON();
        
        now += 60000;
        history.push(createMessage(now, "player2", "Amet"));

        act(() => {
            socket.emit("sendMessageAll", structuredClone(history))
        })

        let final = component.toJSON();
        expect(init).toMatchDiffSnapshot(final);
    });

});
