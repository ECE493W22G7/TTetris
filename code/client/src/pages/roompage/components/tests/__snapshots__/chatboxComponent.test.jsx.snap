// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatboxComponent renders correctly 1`] = `
<div
  className="chatbox-content card"
>
  <div
    className="chat-title font-italic text-center card-header"
  >
    Chat Box
  </div>
  <div
    className="messages-container card"
  >
    <div
      className="container"
    >
      <span
        className="container-name-left"
      >
        player1
      </span>
      <p
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        Lorem
      </p>
      <span
        className="time-right"
      >
        1649451049840
      </span>
    </div>
    <div
      className="container"
    >
      <span
        className="container-name-left"
      >
        player2
      </span>
      <p
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        Ipsum
      </p>
      <span
        className="time-right"
      >
        1649451050840
      </span>
    </div>
    <div
      className="container"
    >
      <span
        className="container-name-left"
      >
        player3
      </span>
      <p
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        Dolor
      </p>
      <span
        className="time-right"
      >
        1649451110840
      </span>
    </div>
  </div>
  <div
    className="send-box"
  >
    <div
      className="input-group mb-3"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Type"
        className="form-control"
        data-testid="send-box-input"
        onChange={[Function]}
        onKeyDownCapture={[Function]}
        placeholder="Type..."
        type="text"
        value=""
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn send-button"
          onClick={[Function]}
          type="button"
        >
          Send
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ChatboxComponent should show message when a socket event is recieved 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -76,10 +76,33 @@
          className=\\"time-right\\"
        >
          1649451232840
        </span>
      </div>
+     <div
+       className=\\"container\\"
+     >
+       <span
+         className=\\"container-name-left\\"
+       >
+         player2
+       </span>
+       <p
+         style={
+           Object {
+             \\"color\\": \\"#fff\\",
+           }
+         }
+       >
+         Amet
+       </p>
+       <span
+         className=\\"time-right\\"
+       >
+         1649451292840
+       </span>
+     </div>
    </div>
    <div
      className=\\"send-box\\"
    >
      <div"
`;
